### PR TITLE
Add new talk voting tests.

### PR DESCRIPTION
**Pull request by umgelurgel**

### CODE REVIEW

File: tests/common_tools.py [hunk 1]
```diff
@@ -203,15 +210,25 @@ def get_default_conference(end=None):
...
+        voting_end=voting_end,
     )
 
     return conference
 
 
-def create_talk_for_user(user):
-    talk = TalkFactory(status=TALK_STATUS.proposed, created_by=user)
+def create_talk_for_user(user, **kwargs):
+    if user == None:
```
Comment (matrixise) [inline tests/common_tools.py]:
```
if user is None:
```

File: docker-compose.yml [hunk 2]
```diff
@@ -10,8 +10,8 @@ services:
       PYTHONDONTWRITEBYTECODE: 1
       DJANGO_SETTINGS_MODULE: "pycon.dev_settings"
       SECRET_KEY: "secret"
-    command: ["./manage.py migrate --no-input && ./manage.py runserver 0.0.0.0:8000"]
+    command: ["./manage.py migrate --no-input && ./manage.py runserver 0.0.0.0:8888"]
     volumes:
       - .:/code
     ports:
-      - "8000:8000"
+      - "8888:8888"
```
Comment (matrixise) [inline docker-compose.yml:17]: I am not aware about that, but why did you change the port?

File: tests/common_tools.py [hunk 3]
```diff
@@ -190,10 +191,16 @@ def create_valid_ticket_for_user_and_fare(user, fare=None):
     return ticket
 
 
-def get_default_conference(end=None):
+def get_default_conference(end=None, voting_start=None, voting_end=None):
     if not end:
```
Comment (matrixise) [inline tests/common_tools.py]: Good job, but what do think about a ConferenceFactory with [django_get_or_create](https://factoryboy.readthedocs.io/en/latest/orms.html#factory.django.DjangoOptions.django_get_or_create)

Example:

```python
class ConferenceFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = 'app.conference'
        django_get_or_create = ('name',)

    name = 'EuroPython 2020'
    code = settings.CONFERENCE_CONFERENCE

    conference_start = factory.Faker('future_date', end_date="+30d")
    conference_end = factory.LazyAttribute(lambda obj: obj.conference_start + timedelta(days=7))
    cfp_start = factory.LazyAttribute(lambda obj: obj.conference_start - timedelta(days=33))
    cfp_stop = factory.LazyAttribute(lambda obj: obj.cfp_start + timedelta(days=3))
    voting_start = factory.LazyAttribute(lambda obj: obj.cfp_start + timedelta(days=3))
    voting_end = factory.LazyAttribute(lambda obj: obj.voting_start + timedelta(days=3))
``` 

you could use it like that:

```python
def get_default_conference(end=None, voting_start=None, voting_end=None):
    parameters = {
        'name': 'EuroPython 2020',
        'voting_start': voting_start,
        'voting_end': voting_end,
        'end': end,
    }
    return ConferenceFactory(**{k: v for k, v in parameters.items() if v is not None})
```

File: tests/test_new_talk_voting.py [hunk 4]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
     response = user_client.get(url)
 
     assert response.status_code == 200
+    assert template_used(response, "ep19/bs/talk_voting/voting.html")
+
+
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_lists_proposed_talks(mock_allowed_to_vote, user_client):
+    get_default_conference()
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:71]: here, we don't use the conference instance, why?

File: tests/test_new_talk_voting.py [hunk 5]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
...
+    assert template_used(response, "ep19/bs/talk_voting/voting.html")
+    assert talk.title in response.content.decode()
+    assert training.title in response.content.decode()
+    assert poster.title in response.content.decode()
+    assert helpdesk.title in response.content.decode()
+
+
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_vote_filters(mock_allowed_to_vote, user_client):
+    get_default_conference()
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:90]: ditto

File: tests/test_new_talk_voting.py [hunk 6]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
...
+
+    assert not_voted.title not in response.content.decode()
+    assert voted.title not in response.content.decode()
+    assert user_talk.title in response.content.decode()
+    assert speaker_talk.title in response.content.decode()
+
+
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_type_filters(mock_allowed_to_vote, user_client):
+    get_default_conference()
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:136]: ditto

File: tests/test_new_talk_voting.py [hunk 7]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
...
+    assert poster.title not in response.content.decode()
+    assert helpdesk.title in response.content.decode()
 
 
-@pytest.mark.xfail
-def test_talk_voting_both_filters():
-    assert False
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_both_filters(mock_allowed_to_vote, user_client):
+    get_default_conference()
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:176]: ditto

File: tests/test_new_talk_voting.py [hunk 8]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
...
-def test_talk_voting_hides_approved_talks():
-    assert False
+    response = user_client.get(url, data={'talk_type': 'talk', 'filter': 'not-voted'})
+
+    assert not_voted_talk.title in response.content.decode()
+    assert voted_training.title not in response.content.decode()
+
+
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_hides_admin_talks(mock_allowed_to_vote, user_client):
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:201]: ditto about the conference instance

File: tests/test_new_talk_voting.py [hunk 9]
```diff
@@ -37,31 +63,164 @@ def test_talk_voting_available_with_proposal(user_client):
...
+    response = user_client.get(url)
+
+    assert response.status_code == 200
+    assert template_used(response, "ep19/bs/talk_voting/voting.html")
+    assert talk.title not in response.content.decode()
+
+
+@mock.patch('conference.talk_voting.is_user_allowed_to_vote', return_value=True)
+def test_talk_voting_hides_accepted_talks(mock_allowed_to_vote, user_client):
+    get_default_conference()
```
Comment (matrixise) [inline tests/test_new_talk_voting.py:215]: ditto

File: tests/test_new_talk_voting.py (hunk 4 above)
Comment (umgelurgel) [reply tests/test_new_talk_voting.py:71]: Most views depend on a conference existing.

File: docker-compose.yml (hunk 2 above)
Comment (umgelurgel) [reply docker-compose.yml:17]: 8000 is the default for a lot of applications - meaning if you work on another project, you'll likely need to bring the other containers down in order to start epcon. This should reduce the chance of a port collision with other projects.

File: tests/common_tools.py (hunk 1 above)
Comment (umgelurgel) [reply tests/common_tools.py]: Thanks, good catch!

File: tests/common_tools.py (hunk 3 above)
Comment (umgelurgel) [reply tests/common_tools.py]: That's a good idea, thanks! going to update that.

Comment (umgelurgel) [reply tests/common_tools.py]: Simplified it a it - stopped short of retiring the method completely and replacing it with a fixture, that's out of scope of this PR.

File: docker-compose.yml (hunk 2 above)
Comment (artcz) [reply docker-compose.yml:17]: Makefile uses 37266 for the same exact reason (port collision)
(37266 is epcon in t9)

File: tests/factories.py [hunk 10]
```diff
@@ -243,32 +243,17 @@ class Meta:
 class ConferenceFactory(factory.django.DjangoModelFactory):
     class Meta:
         model = "conference.Conference"
+        django_get_or_create = ('name',)
```
Comment (artcz) [inline tests/factories.py]: doesn't make much difference but I'd go with get_or_create on code instead, because it's specified in settings and iirc conference code appears in a couple of places in db w/o a FK to Conference.

File: tests/factories.py [hunk 11]
```diff
@@ -243,32 +243,17 @@ class Meta:
...
-        lambda conf: conf.cfp_end + timedelta(days=10)
-    )
-    voting_end = factory.LazyAttribute(
-        lambda conf: conf.voting_start + timedelta(days=5)
-    )
+    conference_start = factory.LazyAttribute(lambda obj: timezone.now() + timedelta(days=30))
+    conference_end = factory.LazyAttribute(lambda obj: obj.conference_start + timedelta(days=5))
+    cfp_start = factory.LazyAttribute(lambda obj: obj.conference_start - timedelta(days=33))
+    cfp_end = factory.LazyAttribute(lambda obj: obj.cfp_start + timedelta(days=6))
+    voting_start = factory.LazyAttribute(lambda obj: obj.cfp_start)
```
Comment (artcz) [inline tests/factories.py:257]: given that almost by definition voting_start should be after cfp_end, I'd leave the comment about that it's specifically set to make testing easier.

File: tests/test_new_talk_voting.py [hunk 12]
```diff
@@ -2,31 +2,57 @@
...
-def test_talk_voting_unavailable_without_a_ticket():
-    assert False
+def test_talk_voting_unavailable_without_a_ticket(user_client):
+    get_default_conference()
+    url = reverse('talk_voting:talks')
+
+    response = user_client.get(url)
+
+    assert response.status_code == 200
+    assert not user_client.user.ticket_set.exists()
```
Comment (artcz) [inline tests/test_new_talk_voting.py:43]: I'd go for explicit `is False` here.

File: tests/test_new_talk_voting.py [hunk 13]
```diff
@@ -143,7 +302,7 @@ def test_dont_publish_talks_without_speaker_details(db, user_client, mocker):
     assert talk2.title in response.content.decode("utf8")
 
 
-def test_dont_publish_talks_without_speaker_details1(db, user_client, mocker):
+def test_dont_publish_talks_without_speaker_details1(db, user_client):
```
Comment (artcz) [inline tests/test_new_talk_voting.py]: why the `1`?

**Review by artcz — APPROVED** [review]
Added some comments, otherwise LGTM

File: docker-compose.yml (hunk 2 above)
Comment (umgelurgel) [reply docker-compose.yml:17]: It used to use that, before the move to docker. I don't mind changing it back to 37266, but it should not stop this PR from being merged.

File: tests/factories.py (hunk 10 above)
Comment (umgelurgel) [reply tests/factories.py]: That's a good idea - even more so as `code` is the PK

File: tests/test_new_talk_voting.py (hunk 13 above)
Comment (umgelurgel) [reply tests/test_new_talk_voting.py]: Must have come from an earlier MR; looks like there's two tests with near identical names, one testing the view, the other testing the queryset. Will update names accordingly.

File: tests/test_new_talk_voting.py (hunk 12 above)
Comment (umgelurgel) [reply tests/test_new_talk_voting.py:43]: I think it's a stylistic preference; I prefer `not [...].exists()` as its reads closer to natural language than  `[..].exists() is False`

File: tests/factories.py (hunk 11 above)
Comment (umgelurgel) [reply tests/factories.py:257]: Good point, will leave a comment.
